File: solutions/cpp/0210.cpp
class Solution {
 public:
  vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
    deque<int> ans;
    vector<vector<int>> graph(numCourses);
    vector<int> state(numCourses, 0);  // 0: init, 1: visiting, 2: visited

    for (const vector<int>& p : prerequisites)
      graph[p[1]].push_back(p[0]);

    function<bool(int)> hasCycle = [&](int course) {
      if (state[course] == 1) return true;
      if (state[course] == 2) return false;

      state[course] = 1;
      for (int neighbor : graph[course])
        if (hasCycle(neighbor))
          return true;
      state[course] = 2;
      ans.push_front(course);

      return false;
    };

    for (int course = 0; course < numCourses; ++course)
      if (hasCycle(course))
        return {};

    return {begin(ans), end(ans)};
  }
};

File: solutions/java/0210.java
class Solution {
  public int[] findOrder(int numCourses, int[][] prerequisites) {
    Deque<Integer> ans = new ArrayDeque<>();
    List<Integer>[] graph = new List[numCourses];
    for (int course = 0; course < numCourses; ++course)
      graph[course] = new ArrayList<>();

    int[] state = new int[numCourses];

    for (final int[] p : prerequisites)
      graph[p[1]].add(p[0]);

    for (int course = 0; course < numCourses; ++course)
      if (hasCycle(course, graph, state, ans))
        return new int[] {};

    return ans.stream().mapToInt(i -> i).toArray();
  }

  boolean hasCycle(int course, List<Integer>[] graph, int[] state, Deque<Integer> ans) {
    if (state[course] == 1)
      return true;
    if (state[course] == 2)
      return false;

    state[course] = 1;
    for (int neighbor : graph[course])
      if (hasCycle(neighbor, graph, state, ans))
        return true;
    state[course] = 2;
    ans.addFirst(course);

    return false;
  }
}